pd-client/client.go

```diff
@@ -14,106 +14,320 @@
+				continue
+			}
+			clusterID, err := client.GetClusterID()
+			if err != nil {
+				log.Errorf("[pd] failed to get cluster id: %v", err)
+				continue
+			}
+			c.clusterID = clusterID
+			return nil
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	return errors.New("failed to get cluster id")
 }
 
-func (c *client) GetClusterID() uint64 {
-	return c.worker.clusterID
+func (c *client) updateLeader() error {
+	for _, u := range c.urls {
+		client, err := apiutil.NewClient(u, pdTimeout)
+		if err != nil {
+			continue
+		}
+		leader, err := client.GetLeader()
+		if err != nil {
+			continue
+		}
+		c.mu.RLock()
+		changed := c.leader != leader.GetAddr()
+		c.mu.RUnlock()
+		if changed {
+			if err = c.switchLeader(leader.GetAddr()); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	}
+	return errors.Errorf("failed to get leader from %v", c.urls)
 }
 
-func (c *client) GetTS() (int64, int64, error) {
-	req := &tsoRequest{
-		done: make(chan error, 1),
+func (c *client) switchLeader(addr string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	log.Infof("[pd] leader switches to: %v, previous: %v", addr, c.leader)
+	if _, ok := c.clients[addr]; !ok {
+		cc, err := grpc.Dial(addr, grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+			u, err := url.Parse(addr)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			return net.Dial(u.Scheme, u.Host)
+		}), grpc.WithInsecure())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		c.clients[addr] = pdpb2.NewPDClient(cc)
 	}
+	c.leader = addr
+	return nil
+}
+
+func (c *client) leaderLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-c.checkLeaderCh:
+		case <-time.After(time.Minute):
+		case <-c.quit:
+			return
+		}
 
+		if err := c.updateLeader(); err != nil {
+			log.Errorf("[pd] failed updateLeader: %v", err)
+		}
+	}
+}
+
+func (c *client) tsLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case first := <-c.tsoRequests:
+			c.procTSORequests(first)
+		case <-c.quit:
+			return
+		}
+	}
+}
+
+func (c *client) procTSORequests(first *tsoRequest) {
 	start := time.Now()
-	c.worker.requests <- req
-	err := <-req.done
+	ctx, cancel := context.WithTimeout(context.Background(), pdTimeout)
+
+	pendingCount := len(c.tsoRequests)
+	resp, err := c.leaderClient().Tso(ctx, &pdpb2.TsoRequest{
+		Header: &pdpb2.RequestHeader{ClusterId: c.clusterID},
+		Count:  uint32(pendingCount + 1),
+	})
+	cancel()
 	requestDuration.WithLabelValues("tso").Observe(time.Since(start).Seconds())
+	if err == nil && resp.GetCount() != uint32(pendingCount+1) {
+		err = errTSOLength
+	}
+	if err != nil {
+		c.finishTSORequest(first, 0, 0, errors.Trace(err))
+		for i := 0; i < pendingCount; i++ {
+			c.finishTSORequest(<-c.tsoRequests, 0, 0, errors.Trace(err))
+		}
+		return
+	}
+
+	physical, logical := resp.GetTimestamp().GetPhysical(), resp.GetTimestamp().GetLogical()
+	c.finishTSORequest(first, physical, logical, nil)
```

Comment:
is it a bug? Seem the first request should have the smallest logic time. 

@huachaohuang